### DISCUSSION TITLE
Be able to set both tags triggers at the same time as branch filtering

### DISCUSSION THREAD

**Discussion by guplem** [codemagic-ci-cd/codemagic-docs#2264] [Ideas]
Workflows Triggered Incorrectly on Tag Creation Across Multiple Branches

**Description:**
When creating tags in a repository with multiple workflows set up for different branches, both workflows are incorrectly triggered even if the tag is created in one specific branch.
The expected behavior is for only the workflow associated with the branch where the tag is created to be triggered.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

**1. Create two workflows for an APP**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/bb619142-2739-4a71-9f1c-3d94a74b6ae5)

**2. Configure them with the same tag trigger but distinct branches ("dev" and "main"):**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/04634f14-7f2a-4adb-ac5f-5f795f790d50)
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/fbfd1a6d-0315-49c5-944a-372f89d8972c)

**3. Push changes and create a new tag to one branch (for instance: "dev"):**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/732b0f6e-e671-4ba6-be82-f6c9250f76ef)

**4. See how both workflows get triggered:**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/c75db3f0-7799-4741-bfbc-da7a347285bc)

Images of the triggers:
**DEV**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/75f90901-a610-42e5-906e-f5c8f9f260ff)
**PRODUCTION**
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/11029629/91676975-f4e2-4179-8d0b-18e99d424b48)

> I expect the "trigger on tag creation" configuration to only be triggered if the "watched branch patterns" and the "watched tag patterns" are matched at the same time.

### Expected results

Only the "Dev" workflow should be triggered.

### Actual results

Both, the "Production" and "Dev" workflows are triggered:

### Build id (optional)

6579cd5124d0a653beeb0d97

**Comment by dtrdic** [comment]
Hello @guplem, based on the Codemagic documentation that's expected behaviour. Check out [this part](https://docs.codemagic.io/flutter-running-builds/starting-builds-automatically/#build-triggers:~:text=Note%20that%20the%20watched%20branch%20settings%20do%20not%20affect%20tag%20builds)

`Note that the watched branch settings do not affect tag builds.`

**Reply by guplem** [reply]
In that case, how can I only trigger builds with on tag creations while only watching a particular branch?
Or... if it is not possible, is there any workaround?

**Reply by dtrdic** [reply]
Unfortunately, there's no workaround for this. If both workflows are configured to trigger on tag creation, they will initiate regardless of the branch

**Reply by guplem** [reply]
Okay, I've created [an issue](https://github.com/codemagic-ci-cd/cli-tools/issues/384) to address this.

**Reply by guplem** [reply]
Since they have closed my issue, I am turning this discussion to an idea as disclosed in [the issue's comment](https://github.com/codemagic-ci-cd/cli-tools/issues/384#issuecomment-1864195528).

**Reply by dtrdic** [reply]
hey @guplem, yes for workfow editor is not possible to do it so this classifies as a feature request.
But actually, If you switch to codemagic.yaml this action is possible, so this could be your workaround. You will just have to define the triggering section like this for each workflow;

```
    triggering:
      events:
          - tag
      tag_patterns:
        - pattern: "*" 
          include: true
      branch_patterns:
      - pattern: main
        include: true
```

```
    triggering:
      events:
          - tag
      tag_patterns:
        - pattern: "*" 
          include: true
      branch_patterns:
      - pattern: dev
        include: true
```

just tried myself and it's working

**Reply by Ycoi** [reply]
Thanks for the idea @guplem! I’ll record this down and we’ll consider this when making other changes to the workflow editor. 

As David said, if you are in a hurry, you can already do this if you switch to the yaml configuration which offers more flexibility also in other ways.  In the long run we are looking into moving to the direction of improved onboarding and visual yaml editor instead of the current workflow editor. This to enable more of our users easy access to advanced options like this. So even though we will keep on maintaining and supporting the workflow editor, the yaml configuration is the more flexible and futureproof option.
